packages/shared/src/components/buttons/ButtonV2.tsx

```diff
@@ -0,0 +1,256 @@
+
+import { ColorName as ButtonColor } from '../../styles/colors';
+
+export enum ButtonSize {
+  XLarge = 'xlarge', // only used for iconOnly buttons
+  Large = 'large',
+  Medium = 'medium',
+  Small = 'small',
+  XSmall = 'xsmall',
+}
+
+const buttonSizeToIconSize: Record<ButtonSize, IconSize> = {
+  [ButtonSize.XLarge]: IconSize.XLarge,
+  [ButtonSize.Large]: IconSize.Large,
+  [ButtonSize.Medium]: IconSize.Medium,
+  [ButtonSize.Small]: IconSize.Small,
+  [ButtonSize.XSmall]: IconSize.XSmall,
+};
+
+export type IconType = React.ReactElement<IconProps>;
+
+export { ButtonColor };
+
+export enum ButtonVariant {
+  Primary = 'primary',
+  Secondary = 'secondary',
+  Tertiary = 'tertiary',
+  Float = 'tertiaryFloat',
+}
+
+export enum ButtonIconPosition {
+  Left = 'left',
+  Right = 'right',
+}
+interface CommonButtonProps {
+  size?: ButtonSize;
+  loading?: boolean;
+  pressed?: boolean;
+  disabled?: boolean;
+  children?: ReactNode;
+  tag?: React.ElementType & AllowedTags;
+}
+
+// when color is present, variant is required
+type ColorButtonProps =
+  | { color: ButtonColor; variant: ButtonVariant }
+  | { color?: never; variant?: ButtonVariant };
+
+// when iconPosition is present, icon is required
+type IconButtonProps =
+  | { iconPosition: ButtonIconPosition; icon: IconType }
+  | { iconPosition?: never; icon?: IconType };
+
+type BaseButtonProps = CommonButtonProps & ColorButtonProps & IconButtonProps;
+
+const useGetIconWithSize = (
+  size: ButtonSize,
+  iconOnly: boolean,
+  iconPosition: ButtonIconPosition,
+) => {
+  return (icon: React.ReactElement<IconProps>) =>
+    React.cloneElement(icon, {
+      size: icon.props?.size ?? buttonSizeToIconSize[size],
+      className: classNames(
+        icon.props.className,
+        !iconOnly && 'text-base !w-6 !h-6',
+        !iconOnly && iconPosition === ButtonIconPosition.Left && '-ml-2 mr-1',
+        !iconOnly && iconPosition === ButtonIconPosition.Right && 'ml-1 -mr-2',
+        // add margin to social icons
+        icon.props.className?.split(' ').includes('socialIcon') && 'mr-3',
+      ),
+    });
+};
+
+export type AllowedTags = keyof Pick<JSX.IntrinsicElements, 'a' | 'button'>;
+export type AllowedElements = HTMLButtonElement | HTMLAnchorElement;
+export type ButtonElementType<Tag extends AllowedTags> = Tag extends 'a'
+  ? HTMLAnchorElement
+  : HTMLButtonElement;
+
+export type ButtonProps<T extends AllowedTags> = BaseButtonProps &
+  HTMLAttributes<T> &
+  JSX.IntrinsicElements[T] & {
+    ref?: Ref<ButtonElementType<T>>;
+  };
+
+const sizeToClassName: Record<ButtonSize, string> = {
+  [ButtonSize.XLarge]: 'border-none',
+  [ButtonSize.Large]: 'h-12 px-6 rounded-14',
+  [ButtonSize.Medium]: 'h-10 px-5 rounded-xl',
+  [ButtonSize.Small]: 'h-8 px-3 rounded-10',
+  [ButtonSize.XSmall]: 'h-6 px-2 rounded-lg',
+};
+
+const iconOnlySizeToClassName: Record<ButtonSize, string> = {
+  [ButtonSize.XLarge]: 'h-16 w-16 p-0 rounded-[1.375rem]',
```

Comment:
What amount is this?
We need to check if we support it maybe, or else add it :)

Reply:
We don't have this at the moment. I was on the fence about adding it vs just using a custom value here, since we do not use it anywhere else right now. But I'll add it 👍

Reply:
Thanks, it's 22px? Yeah let's add it, as we might use it for profiles or images in the future.